### DISCUSSION TITLE
SEN12MS baseline

### DISCUSSION THREAD

**Discussion by taeil** [Berkeley-Data/hpt#47] [Show and tell]
[repo](https://github.com/Berkeley-Data/SEN12MS/tree/taeil)
[wandb](https://wandb.ai/cal-capstone/SEN12MS)

pretrained models are [downloaded](https://syncandshare.lrz.de/getlink/fiCDbqiiSFSNwot5exvUcW1y/trained_models) and placed on s3 (s3://sen12ms/pretrained_sup) for faster download along with other pre-trained models. 

1. use the pretrained models 
2. run various prediction and explore different metrics for baselines 

Once this is done, please convert this issue to the discussion for documentation purpose.

**Comment by suryagutta** [comment]
**Setup Tensorboard**
Code is using tensorboard. It takes logs directory as input to write the tensorboard logs to be retrieved by the tensorboard server to display the metrics. Launched the tensorboard container using the following steps:
1) Run the tensorbpard docker container to avoid the installation of tensorboard packages on the main OS.
docker run --name tnsb -d --rm -p 6006:6006 -v /your/log/dir:/root/logs tensorflow/tensorflow:latest tensorboard --bind_all --logdir /root/logs
2) Access from the browser **http://44.231.248.158:6006/**

**Comment by suryagutta** [comment]
**Train the model**

1. cd ~/SEN12MS
2. docker run --name tf2 --gpus all -it --rm -p 8888:8888 -v $PWD:/workspace/app -v /tmp:/tmp -v $PWD/../data/processed:/workspace/app/data tf2
3. cd classification
4. pip install -r requirements.txt
5. Copy IGBP_probability_labels.pkl from labels to splits folder as main_train.py is looking for it under splits folder  
cp /workspace/app/labels/IGBP_probability_labels.pkl /workspace/app/splits/IGBP_probability_labels.pkl  
7. CUDA_VISIBLE_DEVICES=0 python main_train.py --exp_name sem12ms_baseline --data_dir /workspace/app/data/sen12ms --label_split_dir /workspace/app/splits --use_RGB --IGBP_simple --label_type multi_label --threshold 0.1 --model DenseNet121 --lr 0.001 --decay 1e-5 --batch_size 64 --num_workers 4 --epochs 1

Initially, it failed as there is no Validation split file. It was not provided and they expect us to extract the validation data directly from the train split file. To extract, either we need to use the train_list.pkl file (code uses pkl) or a train_list.txt file. Will modify the code to take a sample from the train_list.pkl and write to **val_list.pkl** .

There are a total of `162555` entries in current train_list.txt file are there. Did a random split into `88.6%:11.4%` (`test_list_updated.txt`: `val_list_updated.txt`). Chose this ratio as in dataset.py, there is a comment # `18550` samples in val set. 

The following are the counts:
- train_list_updated.txt/.pkl  - 144024
- val_list_updated.txt/.pkl  - 18532

Also, the current test dataset is also around 10~11 percent of the total dataset.

Modified the classification/dataset.py (took backup as dataset_original.py) to use train_list_updated.pkl and val_list_updated.pkl. 

**main_train.py is failing with rasterio._err.CPLE_OpenFailedError:** /workspace/app/data/sen12ms/ROIs1158_spring/s2_101/ROIs1158_spring_s2_101_p138.tif : No such file or directory
The above file exists. There are no permission issues. Did a google search and found that sometimes it fails due to relative paths. It might also fail due to rasterio(gdal) issues. **Need to investigate.**

To test the above issue, I ran the following in python console inside the container and it worked fine.
```
import rasterio
with rasterio.open('/workspace/app/data/sen12ms/ROIs1158_spring/s2_101/ROIs1158_spring_s2_101_p138.tif') as src:
    print(src.width, src.height)
    print(src.crs)
    print(src.transform)
    print(src.count)
    print(src.indexes)
```

**Comment by suryagutta** [comment]
Found the rootcause. The downloaded train_list.txt has '\n' characters at the end of each line. Trimmed those using rstrip('\n') and regenerated the pickle files. It started working!

**Comment by suryagutta** [comment]
**For 1 epochs:**
Validation microPrec: 0.693667 microF1: 0.539214 sampleF1: 0.495452 microF2: 0.475667 sampleF2: 0.479950
I Will work on improving. Also, need to fix the tensorboard log directory path.

**Comment by suryagutta** [comment]
Baseline to be reproduced
![image](https://user-images.githubusercontent.com/15736461/109337210-51fd4380-7819-11eb-868b-789f1baa3944.png)

**Comment by taeil** [comment]
I will change to wandb instead of tensorboard today or tomorrow.

**Comment by suryagutta** [comment]
**Baseline evaluation based on the Paper**
**First network (image classification using ResNet-110)** 
**Input:** Summer subset with cut images 64x64 from S2 samples (10 bands)
**Test data:** City of Munich

**Second network (image segmentation using DenseNet)**
**Input:** S2 samples with full size images 256x256 (10 bands)
**Test data:** City of Rome

**Comment by taeil** [comment]
[running 50 epochs ](https://wandb.ai/cal-capstone/SEN12MS/table?workspace=user-taeil)

[code change](https://github.com/Berkeley-Data/SEN12MS/tree/taeil)

**Comment by taeil** [comment]
https://wandb.ai/cal-capstone/SEN12MS

**Comment by taeil** [comment]
started new evaluation on the baseline pre-training models (supervised) 

- [single_label/single_DenseNet121_s1s2](https://wandb.ai/cal-capstone/SEN12MS/runs/2rmv5iy3)
- [single_label/single_DenseNet121_s2](https://wandb.ai/cal-capstone/SEN12MS/runs/14gp4x2p) 
- [single_label/single_ResNet50_RGB ](https://wandb.ai/cal-capstone/SEN12MS/runs/h0ss5sbi)
- single_label/single_ResNet50_s1s2
- (more)

**Comment by tchken** [comment]
single_label/single_DenseNet121_s1s2
single_label/single_DenseNet121_s2
single_label/single_ResNet50_RGB


the link is showing blank. can I check if this is the right link? thanks

**Comment by tchken** [comment]
- https://wandb.ai/cal-capstone/SEN12MS/table?workspace=user-taeil
- For both baseline and CLS training evaluation (from my perspective)
1. Must tracked metrics (loss, mircro f1, macro f1, top-1 accuracy, top-1 precision, top-1 recall, class-wise accuracy. epochs, batch size, lr, label_type, model, use_{s1, s2} )
2. To be checked with Colorado that whether these needs to be reported ( top - {2...}acc, f{2..}, overall acc, average acc, use_RGB)

Thanks

**Comment by tchken** [comment]
- collected current results from w&b for presentations purpose. As discussed, this serves as the staring points for evaluation purposes, likely will revise to be completed.

![Screen Shot 2021-03-16 at 6 46 50 PM](https://user-images.githubusercontent.com/39634122/111389931-36eb5a00-8688-11eb-8d5c-c060d2f9b298.png)

- when we have the results from the new pre-trained embedding, I will compare and write up more completed insights in the paper (pending). similar but not limited to below reference.
<img width="289" alt="Screen Shot 2021-03-16 at 6 51 02 PM" src="https://user-images.githubusercontent.com/39634122/111390109-90538900-8688-11eb-8b01-a33359768020.png">
